**Импорт**

In [1]:
# Библиотека для работы с табличными данными
import pandas as pd

**Чтение и предварительный просмотр данных о Титанике**

In [5]:
df = pd.read_csv('sample_data/titanic.csv')

In [6]:
# Посмотреть первые 5 строк 
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [20]:
# Посмотреть последние 5 строк
df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [8]:
df.shape

(891, 12)

Размер нашей таблицы: **12 стоблцов** на **891 строку**

In [9]:
# Посмотреть есть ли у нас пустые значения
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Таким образом, у нас отсутствуют следующие данные:
- не извествен возраст 177 пассажиров;
- не известно в каких каютах были размещены 687 человек;
- не известно из какого порта на борт Титаника попало 2 человека.

In [12]:
# Информация о нашем файле
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


 В представленном файле имееютя следующе данные:
 1. **PassengerId** - идентификатор пассажира;
 2. **Survived** 	- выжившие;
 3. **Pclass** 	- класс пассажира;
 4. **Name** - Ф.И.О. пассажира;
 5. **Sex** - пол пассажира;
 6. **Age** - возраст пассажира;
 7. **SibSp** — число братьев, сестер или супругов на борту у человека;
 8. **Parch** — количество родителей, детей (в том числе приемных);
 9. **Ticket** - билет;
 10. **Fare** - плата за проезд;
 11. **Cabin** -  каюта;
 12. **Embarked** — порт посадки.

**Количество мужчин и женщин на борту**

In [109]:
# Количество мужчин и женщин на борту
df.groupby('Sex')['PassengerId'].count()

Sex
female    314
male      577
Name: PassengerId, dtype: int64

**Возраст пассажиров:**

In [113]:
print('Максимальный возраст: ',end='')
print(df['Age'].max())
print('Минимальный возраст: ',end='')
print(df['Age'].min())
print('Средний возраст: ',end='')
print(df['Age'].mean())

Максимальный возраст: 80.0
Минимальный возраст: 0.42
Средний возраст: 29.69911764705882


**Данные о самом старом пассжире**

In [47]:
# Ф.И.О. самого старого пассажира
df[['Name', 'Age']].max()

Name    van Melkebeke, Mr. Philemon
Age                            80.0
dtype: object

**Данные о самом молодом пассажире**

In [49]:
# Ф.И.О. самого молодого пассажира
df[['Name', 'Age']].min()

Name    Abbing, Mr. Anthony
Age                    0.42
dtype: object

**Количество выживших мужчин и женщин**

In [125]:
# Количество выживших пасажиров
df.groupby('Sex')['Survived'].sum()


Sex
female    233
male      109
Name: Survived, dtype: int64

In [82]:
df.loc[(df['Survived']) & (df['Age']<18) & (df['Sex'])]


PassengerId  Survived  Pclass                                      Name  \
9             10         1       2       Nasser, Mrs. Nicholas (Adele Achem)   
10            11         1       3           Sandstrom, Miss. Marguerite Rut   
22            23         1       3               McGowan, Miss. Anna "Annie"   
39            40         1       3               Nicola-Yarred, Miss. Jamila   
43            44         1       2  Laroche, Miss. Simonne Marie Anne Andree   
..           ...       ...     ...                                       ...   
830          831         1       3   Yasbeck, Mrs. Antoni (Selini Alexander)   
831          832         1       2           Richards, Master. George Sibley   
853          854         1       1                 Lines, Miss. Mary Conover   
869          870         1       3           Johnson, Master. Harold Theodor   
875          876         1       3          Najib, Miss. Adele Kiamie "Jane"   

        Sex    Age  SibSp  Parch         Ticket     Fare Cabin Embarked  
9    female  14.00      1      0         237736  30.0708   NaN        C  
10   female   4.00      1      1        PP 9549  16.7000    G6        S  
22   female  15.00      0      0         330923   8.0292   NaN        Q  
39   female  14.00      1      0           2651  11.2417   NaN        C  
43   female   3.00      1      2  SC/Paris 2123  41.5792   NaN        C  
..      ...    ...    ...    ...            ...      ...   ...      ...  
830  female  15.00      1      0           2659  14.4542   NaN        C  
831    male   0.83      1      1          29106  18.7500   NaN        S  
853  female  16.00      0      1       PC 17592  39.4000   D28        S  
869    male   4.00      1      1         347742  11.1333   NaN        S  
875  female  15.00      0      0           2667   7.2250   NaN        C  

[61 rows x 12 columns]

Спасли **61 пассажира**, чей возраст меньше **18 лет**

**Данные выжившего пассажира с максимальным возрастом**

In [127]:
df.loc[(df['Survived']) & (df['Age']==df['Age'].max())]

PassengerId  Survived  Pclass                                  Name  \
630          631         1       1  Barkworth, Mr. Algernon Henry Wilson   

      Sex   Age  SibSp  Parch Ticket  Fare Cabin Embarked  
630  male  80.0      0      0  27042  30.0   A23        S

In [88]:
# Максимальный цена билета
print(df['Fare'].max())
# Минимальный цена билета
print(df['Fare'].min())
#Средняя цена билета
print(df['Fare'].mean())

512.3292
0.0
32.204207968574636


In [99]:
df[df['Fare'] ==0]

PassengerId  Survived  Pclass                              Name   Sex  \
179          180         0       3               Leonard, Mr. Lionel  male   
263          264         0       1             Harrison, Mr. William  male   
271          272         1       3      Tornquist, Mr. William Henry  male   
277          278         0       2       Parkes, Mr. Francis "Frank"  male   
302          303         0       3   Johnson, Mr. William Cahoone Jr  male   
413          414         0       2    Cunningham, Mr. Alfred Fleming  male   
466          467         0       2             Campbell, Mr. William  male   
481          482         0       2  Frost, Mr. Anthony Wood "Archie"  male   
597          598         0       3               Johnson, Mr. Alfred  male   
633          634         0       1     Parr, Mr. William Henry Marsh  male   
674          675         0       2        Watson, Mr. Ennis Hastings  male   
732          733         0       2              Knight, Mr. Robert J  male   
806          807         0       1            Andrews, Mr. Thomas Jr  male   
815          816         0       1                  Fry, Mr. Richard  male   
822          823         0       1   Reuchlin, Jonkheer. John George  male   

      Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
179  36.0      0      0    LINE   0.0   NaN        S  
263  40.0      0      0  112059   0.0   B94        S  
271  25.0      0      0    LINE   0.0   NaN        S  
277   NaN      0      0  239853   0.0   NaN        S  
302  19.0      0      0    LINE   0.0   NaN        S  
413   NaN      0      0  239853   0.0   NaN        S  
466   NaN      0      0  239853   0.0   NaN        S  
481   NaN      0      0  239854   0.0   NaN        S  
597  49.0      0      0    LINE   0.0   NaN        S  
633   NaN      0      0  112052   0.0   NaN        S  
674   NaN      0      0  239856   0.0   NaN        S  
732   NaN      0      0  239855   0.0   NaN        S  
806  39.0      0      0  112050   0.0   A36        S  
815   NaN      0      0  112058   0.0  B102        S  
822  38.0      0      0   19972   0.0   NaN        S

выше указаны данные пассажиров, которые не платили за билтеы, скорее всего это **экипаж Титаника**

In [100]:
df[df['Fare'] ==(df['Fare'].max())]

PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
737          738         1       1              Lesurer, Mr. Gustave J   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  
258  female  35.0      0      0  PC 17755  512.3292          NaN        C  
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C  
737    male  35.0      0      0  PC 17755  512.3292         B101        C

Данные о пассажирах, которые заплатили самую высокую цену за билеты, при этом просмотрев стоблец **Survived** видно, что эти **пассажиры еще и выжили**!!!

**Класс билетов**

In [128]:
df.groupby('Pclass')['PassengerId'].count()

Pclass
1    216
2    184
3    491
Name: PassengerId, dtype: int64

**Описательная статистика**

In [148]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

**Описательная статискита братьев и сестер**

In [180]:
df.SibSp.describe()

count    891.000000
mean       0.523008
std        1.102743
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        8.000000
Name: SibSp, dtype: float64

Максимальное количество братьев и сестер: **8.**
у 75% пассажиров на борту находился 1 родственник.


**Общее количество родителей на борту**

In [177]:
df.Parch.sum()


340

**Количество пассажиров, выживших после катастрофы и имеющих родственные связи**

In [205]:
df.loc[(df['Survived']==1) & (df['Parch'])& (df['SibSp'])]

PassengerId  Survived  Pclass  \
10            11         1       3   
25            26         1       3   
65            66         1       3   
128          129         1       3   
172          173         1       3   
193          194         1       2   
248          249         1       1   
279          280         1       3   
319          320         1       1   
323          324         1       2   
328          329         1       3   
340          341         1       2   
348          349         1       3   
407          408         1       2   
416          417         1       2   
440          441         1       2   
489          490         1       3   
530          531         1       2   
549          550         1       2   
558          559         1       1   
580          581         1       2   
581          582         1       1   
587          588         1       1   
622          623         1       3   
670          671         1       2   
709          710         1       3   
750          751         1       2   
755          756         1       2   
774          775         1       2   
801          802         1       2   
820          821         1       1   
831          832         1       2   
835          836         1       1   
856          857         1       1   
869          870         1       3   
871          872         1       1   

                                                  Name     Sex    Age  SibSp  \
10                     Sandstrom, Miss. Marguerite Rut  female   4.00      1   
25   Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.00      1   
65                            Moubarek, Master. Gerios    male    NaN      1   
128                                  Peter, Miss. Anna  female    NaN      1   
172                       Johnson, Miss. Eleanor Ileen  female   1.00      1   
193                         Navratil, Master. Michel M    male   3.00      1   
248                      Beckwith, Mr. Richard Leonard    male  37.00      1   
279                   Abbott, Mrs. Stanton (Rosa Hunt)  female  35.00      1   
319  Spedden, Mrs. Frederic Oakley (Margaretta Corn...  female  40.00      1   
323  Caldwell, Mrs. Albert Francis (Sylvia Mae Harb...  female  22.00      1   
328     Goldsmith, Mrs. Frank John (Emily Alice Brown)  female  31.00      1   
340                     Navratil, Master. Edmond Roger    male   2.00      1   
348             Coutts, Master. William Loch "William"    male   3.00      1   
407                     Richards, Master. William Rowe    male   3.00      1   
416    Drew, Mrs. James Vivian (Lulu Thorne Christian)  female  34.00      1   
440        Hart, Mrs. Benjamin (Esther Ada Bloomfield)  female  45.00      1   
489              Coutts, Master. Eden Leslie "Neville"    male   9.00      1   
530                           Quick, Miss. Phyllis May  female   2.00      1   
549                     Davies, Master. John Morgan Jr    male   8.00      1   
558             Taussig, Mrs. Emil (Tillie Mandelbaum)  female  39.00      1   
580                        Christy, Miss. Julie Rachel  female  25.00      1   
581  Thayer, Mrs. John Borland (Marian Longstreth M...  female  39.00      1   
587                   Frolicher-Stehli, Mr. Maxmillian    male  60.00      1   
622                                   Nakid, Mr. Sahid    male  20.00      1   
670  Brown, Mrs. Thomas William Solomon (Elizabeth ...  female  40.00      1   
709  Moubarek, Master. Halim Gonios ("William George")    male    NaN      1   
750                                  Wells, Miss. Joan  female   4.00      1   
755                          Hamalainen, Master. Viljo    male   0.67      1   
774              Hocking, Mrs. Elizabeth (Eliza Needs)  female  54.00      1   
801        Collyer, Mrs. Harvey (Charlotte Annie Tate)  female  31.00      1   
820  Hays, Mrs. Charles Melville (Clara Jennings Gr...  female  52.00      1   
831                    Richards, Maste